velox/exec/HashProbe.cpp

```diff
@@ -543,160 +536,191 @@ void HashProbe::fillFilterInput(vector_size_t size) {
             BaseVector::wrapInConstant(
-                numRows, i, input_->childAt(p.inputChannel));
+                numRows, row, input_->childAt(projection.inputChannel));
       }
       EvalCtx evalCtx(
-          operatorCtx_->execCtx(), filter_.get(), filterInputBuildSide_.get());
-      filter_->eval(filterRowsBuildSide_, evalCtx, filterResultBuildSide_);
-      decodedVectorPerRow_.decode(
-          *filterResultBuildSide_[0], filterRowsBuildSide_);
-      bool passed = !filterRowsBuildSide_.testSelected([&](vector_size_t j) {
-        return decodedVectorPerRow_.isNullAt(j) ||
-            !decodedVectorPerRow_.valueAt<bool>(j);
-      });
+          operatorCtx_->execCtx(), filter_.get(), filterTableInput_.get());
+      filter_->eval(filterTableInputRows_, evalCtx, filterTableResult_);
+      decodedFilterTableResult_.decode(
+          *filterTableResult_[0], filterTableInputRows_);
+      const bool passed =
+          !filterTableInputRows_.testSelected([&](vector_size_t j) {
+            return decodedFilterTableResult_.isNullAt(j) ||
+                !decodedFilterTableResult_.valueAt<bool>(j);
+          });
       if (passed) {
-        rows.setValid(i, false);
+        rows.setValid(row, false);
       }
     });
   }
   rows.updateBounds();
 }
 
-vector_size_t HashProbe::evalFilterInNullAwareAntiJoin(
+vector_size_t HashProbe::evalFilterForNullAwareAntiJoin(
     vector_size_t numRows,
-    const Expr& filter) {
-  auto rawMapping = rowNumberMapping_->asMutable<vector_size_t>();
-  SelectivityVector skipRows(numRows, false);
-  SelectivityVector testNullKeyRows(input_->size(), false);
-  SelectivityVector testAllRows(input_->size(), false);
+    const bool filterPropagateNulls) {
+  auto* rawOutputProbeRowMapping =
+      outputRowMapping_->asMutable<vector_size_t>();
+
+  SelectivityVector nullKeyProbeRows(input_->size(), false);
+  SelectivityVector crossJoinProbeRows(input_->size(), false);
+
   for (auto i = 0; i < numRows; ++i) {
-    auto j = rawMapping[i];
-    if (filter.propagatesNulls() && nullFilterProbeInputRows_.isValid(i)) {
-      skipRows.setValid(i, true);
-    } else if (nonNullRows_.isValid(j)) {
-      if (!decodedFilterResult_.isNullAt(i) &&
-          decodedFilterResult_.valueAt<bool>(i)) {
-        skipRows.setValid(i, true);
-      } else {
-        testNullKeyRows.setValid(j, true);
+    // Skip filter input row if it has any null probe side filter column.
+    if (filterPropagateNulls && nullFilterInputRows_.isValid(i)) {
+      continue;
+    }
+
+    const auto probeRow = rawOutputProbeRowMapping[i];
+    if (nonNullInputRows_.isValid(probeRow)) {
+      if (decodedFilterResult_.isNullAt(i) ||
+          !decodedFilterResult_.valueAt<bool>(i)) {
+        nullKeyProbeRows.setValid(probeRow, true);
       }
-    } else {
-      testAllRows.setValid(j, true);
-    }
-  }
-  skipRows.updateBounds();
-  skipRows.applyToSelected([&](vector_size_t i) {
```

Comment:
This masks out some of the cross join rows in filtering, we should keep this.

Reply:
But those are false by default? And only set if it is not skipped in for loop above?

Reply:
There could be intersection.  In case of intersection, the `skipRows` (filter having results) take precedence